instrumentation/httpurlconnection/src/main/java/com/nr/agent/instrumentation/httpurlconnection/MetricState.java

```diff
@@ -10,90 +10,261 @@
+     *
+     * @param tx        current Transaction
+     * @param operation HttpURLConnection method being invoked
+     */
+    private void startSegmentIfNull(Transaction tx, String operation) {
+        if (segment == null) {
+            segment = tx.startSegment(CATEGORY, operation);
+            segment.setMetricName(CATEGORY, LIBRARY, operation);
+        }
+    }
+
+    /**
+     * Track which HttpURLConnection method is called first.
+     * If it is one of the methods calling nonNetworkPreamble (i.e. connect or getOutputStream), then start a
+     * TimerTask to potentially end the timing of the Segment that was started when either of these methods were called.
+     *
+     * @param connection HttpURLConnection
+     * @param operation  HttpURLConnection method being invoked
+     */
+    private void handleSegmentsForNonNetworkMethods(HttpURLConnection connection, String operation) {
+        /*
+         * It's possible to call connect and then getOutputStream, in which case we reset firstMethodCalled
+         * to getOutputStream so that the timer task properly ends the segment and calls reportAsExternal.
+         */
+        if (firstMethodCalled == null || (firstMethodCalled.equals(CONNECT_OP) && operation.equals(GET_OUTPUT_STREAM_OP))) {
+            firstMethodCalled = operation;
+            if (operation.equals(CONNECT_OP) || operation.equals(GET_OUTPUT_STREAM_OP)) {
+                startSegmentExpirationTimerTask(connection);
+            }
+        }
+    }
+
+    /**
+     * If connect, or some combination of connect and getOutputStream, were the first
+     * and only method(s) invoked from the HttpURLConnection API within a defined
+     * time period then the timer task will the end the segment timing to
+     * avoid hitting the default segment timeout of 10 minutes.
+     * <p>
+     * If the segment_timeout is manually configured to be lower than the timer delay set here
+     * then the segment timing will already have been ended and calling end again here
+     * will have no effect.
+     *
+     * @param connection HttpURLConnection
+     */
+    private void startSegmentExpirationTimerTask(HttpURLConnection connection) {
+        Timer timer = new Timer("HttpURLConnection Segment Expiration Timer");
+        TimerTask task = new TimerTask() {
+            public void run() {
+                endSegmentForNonNetworkCall(connection);
+            }
+        };
+
+        /*
+         * The following tests do a Thread.sleep to account for this delay. If this value is changed then the tests will also need to be updated.
+         * functional_test/src/test/java/com/newrelic/agent/instrumentation/pointcuts/net/HttpURLConnectionTest
+         * instrumentation/httpurlconnection/src/test/java/com/nr/agent/instrumentation/httpurlconnection/MetricStateConnectTest
+         */
+        long segmentExpirationDelayInMillis = 1000L;
```

Comment:
Is 1 second a reasonable default for ending the segment timing when only `connect` and/or `getOutputStream` are called? Or does a different value make more sense?

Reply:
Hmmm, the github test runner might have answered the first part of this for me. I don't think 1 second is long enough but I'm not sure what is ideal.

I'm seeing `com.newrelic.agent.instrumentation.pointcuts.net.HttpURLConnectionTest > connectFirstTest` fail regularly when run here on GHA even though it passes consistently when run locally.

This test uses the `HttpURLConnection` API as follows:

```
connection.connect();
connection.getOutputStream();
connection.connect(); // should be no-op
```

This would be a scenario where the `TimerTask` expires, which should result in reporting the external call and ending the segment timing. However, I believe the reason that it is failing is that the 1 second timeout isn't long enough to evaluate `getOutputStream` once it is invoked after the first `connect` call, which means that the segment timing is simply ended and ignored, causing the test assertion to fail because no `External/` metrics are reported.

Given this, I think it means that a 1 second timeout likely won't be long enough in some customer environments either.

Reply:
I upped the timeout to 10 seconds and the tests passed on GHA first time. I guess that confirms the issue.

New question, is 10 seconds a reasonable default?

Reply:
This should no longer be a concern. The code has been refactored so that the `TimerTask` will only ever apply if only `connect` is called, in which case it will timeout in 1 minute and ignore the segment so that it won't ever show up in a trace.